walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,180 @@
+      return byEndTime;
+  }
+
+  /**
+    * Finds the times in a day when mandatory attendees for a meeting are unavailable.
+    *
+    * @param events the events happening that day
+    * @param mandatory the attendees whose attendance is required at another meeting
+    *
+    * @return an ArrayList of TimeRange signifying when mandatory attendees are at another
+    *         event.
+    */
+  private ArrayList<TimeRange> findBusyTimes(Collection<Event> events, List<String> mandatory) {
+      ArrayList<TimeRange> busy = new ArrayList<TimeRange> ();
+
+      for (Event event : events) {          
+          boolean relevantEvent = false;
+          for (String attendee : event.getAttendees()) {
+              if (mandatory.contains(attendee)) {
+                  relevantEvent = true;
+                  break; // only need confirmation that one mandatory attendee is busy
+              }
+          }
+          if (relevantEvent) {
+              busy.add(event.getWhen());
+          }
+      }
+
+      return busy;
+  }
+
+  /**
+    * Finds the times in a day when mandatory attendees for a meeting are available for a given
+    * duration.
+    *
+    * @param busyTimeRanges the times where at least one attendees is unavailable, sorted by start time
+    * @param busybyEndTime the times where at least one attendees is unavailable, sorted by end time
+    * @param duration the length of the new meeting to be added
+    *
+    * @return an ArrayList of TimeRange signifying when mandatory attendees are available for at least
+    *         time duration.
+    */ 
+  private List<TimeRange> findFreeTimes(List<TimeRange> busyTimeRanges, List<TimeRange> busyByEndTime,
+                                        long duration) {
+      
+      List<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+      int startTimeIndex = 0;
+      int endTimeIndex = 0;
+      
+      int numBusyIntervals = busyTimeRanges.size();
+      int lastBusyTime = busyByEndTime.get(numBusyIntervals-1).end();
+
+      // handle beginning of the day before busy intervals begin
+      TimeRange possibleTime = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, 
+                                                      busyTimeRanges.get(endTimeIndex).start(), false);
+      if(possibleTime.duration() >= duration) {
+            meetingTimes.add(possibleTime);
+      }
+      endTimeIndex++;
+
+      // find all free times between scheduled events
+      while (startTimeIndex < numBusyIntervals && endTimeIndex < numBusyIntervals) {
+          int startTime = busyTimeRanges.get(startTimeIndex).end();
+          int endTime = busyTimeRanges.get(endTimeIndex).start();
+
+          if (startTime > endTime) {
+              endTimeIndex++;
+          } else if (startTime == endTime) {
+              startTimeIndex++;
+              endTimeIndex++;
+          } else {
+              possibleTime = TimeRange.fromStartEnd(startTime, endTime, false);
```

Comment:
I think this approach doesn't pass the case where there are no times because all of the events actually overlap: e.g.
```
     |--------------------B----------------------|
|---A---|  <this gap is longer than duration>  |----C------|
```
I haven't tested this myself, but I think your algorithm would start at the end of A and compare with beginning of C, which is longer than duration, but is also actually covered by B.

IMO The overlapping intervals is not a trivial problem; I believe you cover in the upper division algorithms class haha so don't worry about it.

Reply:
I actually believe that this is covered by one of the test cases and my code passes it! But I will diagram it out on paper and change it if necessary tomorrow.

Reply:
This is the example that I am thinking of:
```
  @Test
  public void overlappingTest() {
    // Events  :       |--A--|     |--B--|
    //                    |-----C-----|
    // Day     : |-----------------------------|
    // Options : |--1--|                 |--2--|

    Collection<Event> events = Arrays.asList(
        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
            Arrays.asList(PERSON_A)),
        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
            Arrays.asList(PERSON_B)),
        new Event("Event 3", TimeRange.fromStartDuration(TimeRange.getTimeInMinutes(8, 15), DURATION_60_MINUTES),
            Arrays.asList(PERSON_C)));

    MeetingRequest request =
        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B, PERSON_C), 30);

    Collection<TimeRange> actual = query.query(events, request);
    Collection<TimeRange> expected =
        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));

    Assert.assertEquals(expected, actual);
  }
```
But  I might have coded up the test incorrectly.  I just copy and modified a similar test from your code :P

 
![Screen Shot 2020-07-09 at 8 12 14 PM](https://user-images.githubusercontent.com/66842831/87112444-a7e89100-c220-11ea-90ac-0911bf70f8c5.png)